pkgs/development/libraries/harfbuzz/default.nix

```diff
@@ -69,7 +71,7 @@ stdenv.mkDerivation (finalAttrs: {
 
   depsBuildBuild = [
     pkg-config
-  ];
+  ] ++ lib.optionals (stdenv.hostPlatform.isMinGW) [ cmake ];
```

Comment:
What's this for?

Reply:
When cross compiling for mingw32. The build is done through cmake.

Reply:
If you don't have this line. You get this error.
```
error: builder for '/nix/store/mbpymd5afiyibnmyk7m305vv4qka3vzq-harfbuzz-i686-w64-mingw32-7.3.0.drv' failed with exit code 1;
       last 10 log lines:
       > Library m found: YES
       > Found pkg-config: /nix/store/gblj7xj9g8j4fma1y7vi4n9dx5q830m8-i686-w64-mingw32-pkg-config-wrapper-0.29.2/bin/i686-w64-mingw32-pkg-config (0.29.2)
       > Run-time dependency freetype2 found: YES 26.1.20
       > Found CMake: NO
       > Run-time dependency glib-2.0 found: NO (tried pkgconfig and cmake)
       > Looking for a fallback subproject for the dependency glib-2.0
       >
       > meson.build:114:11: ERROR: Automatic wrap-based subproject downloading is disabled
       >
       > A full log can be found at /build/harfbuzz-7.3.0/build/meson-logs/meson-log.txt
       For full logs, run 'nix log /nix/store/mbpymd5afiyibnmyk7m305vv4qka3vzq-harfbuzz-i686-w64-mingw32-7.3.0.drv'.

```

Reply:
Please post the full log. The last 10 lines are usualy not enough to get a full picture.

Reply:
Here is the full log. Not sure why some characters look weird.
```
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/hwn45vmbgl2sir2p7ajd4i9knhkw6mil-harfbuzz-7.3.0.tar.xz
source root is harfbuzz-7.3.0
setting SOURCE_DATE_EPOCH to timestamp 1683667985 of file harfbuzz-7.3.0/docs/html/working-with-harfbuzz-clusters.html
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
patching script interpreter paths in src/check-c-linkage-decls.py src/check-externs.py src/check-header-guards.py src/check-includes.py src/check-libstdc++.py src/check-static-inits.py src/check-symbols.py src/fix_get_types.py src/gen-arabic-joining-list.py src/gen-arabic-table.py src/gen-def.py src/gen-emoji-table.py src/gen-harfbuzzcc.py src/gen-hb-version.py src/gen-indic-table.py src/gen-os2-unicode-ranges.py src/gen-ragel-artifacts.py src/gen-tag-table.py src/gen-ucd-table.py src/gen-use-table.py src/gen-vowel-constraints.py test
src/check-c-linkage-decls.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/check-externs.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/check-header-guards.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/check-includes.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/check-libstdc++.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/check-static-inits.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/check-symbols.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/fix_get_types.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-arabic-joining-list.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-arabic-table.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-def.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-emoji-table.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-harfbuzzcc.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-hb-version.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-indic-table.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-os2-unicode-ranges.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-ragel-artifacts.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-tag-table.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-ucd-table.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-use-table.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
src/gen-vowel-constraints.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/subset/run-repack-tests.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/subset/run-tests.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/fuzzing/run-shape-fuzzer-tests.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/fuzzing/run-subset-fuzzer-tests.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/fuzzing/run-draw-fuzzer-tests.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/record-test.sh: interpreter directive changed from "#!/bin/bash" to "/nix/store/q8qq40xg2grfh9ry1d9x4g7lq4ra7n81-bash-5.2-p21/bin/bash"
test/shape/hb-diff: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/hb-diff-stat: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/hb-diff-colorize: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/hb-unicode-encode: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/run-tests.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/hb-unicode-decode: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/data/aots/update.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/data/text-rendering-tests/update.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/hb-diff-filter-failures: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
test/shape/hb-unicode-prettyname: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/bin/python3"
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config.guess
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
mesonConfigurePhase flags: --prefix=/nix/store/g9rp9i1fpxp89s9324a2j38fmzcxbp6s-harfbuzz-i686-w64-mingw32-7.3.0 --libdir=/nix/store/g9rp9i1fpxp89s9324a2j38fmzcxbp6s-harfbuzz-i686-w64-mingw32-7.3.0/lib --libexecdir=/nix/store/g9rp9i1fpxp89s9324a2j38fmzcxbp6s-harfbuzz-i686-w64-mingw32-7.3.0/libexec --bindir=/nix/store/l4qv5ss554zyislxa0fg13r6qrb6f9ps-harfbuzz-i686-w64-mingw32-7.3.0-dev/bin --sbindir=/nix/store/l4qv5ss554zyislxa0fg13r6qrb6f9ps-harfbuzz-i686-w64-mingw32-7.3.0-dev/sbin --includedir=/nix/store/l4qv5ss554zyislxa0fg13r6qrb6f9ps-harfbuzz-i686-w64-mingw32-7.3.0-dev/include --mandir=/nix/store/l4qv5ss554zyislxa0fg13r6qrb6f9ps-harfbuzz-i686-w64-mingw32-7.3.0-dev/share/man --infodir=/nix/store/l4qv5ss554zyislxa0fg13r6qrb6f9ps-harfbuzz-i686-w64-mingw32-7.3.0-dev/share/info --localedir=/nix/store/g9rp9i1fpxp89s9324a2j38fmzcxbp6s-harfbuzz-i686-w64-mingw32-7.3.0/share/locale -Dauto_features=enabled -Dwrap_mode=nodownload --buildtype=plain --cross-file=/nix/store/prcgzd6gqj5bahkpkv7ywjgpny8ypg2z-cross-file.conf -Dcairo=disabled -Dchafa=disabled -Dcoretext=disabled -Dgraphite=disabled -Dicu=disabled -Dintrospection=disabled
[1mThe Meson build system[0m
Version: 1.2.3
Source dir: [1m/build/harfbuzz-7.3.0[0m
Build dir: [1m/build/harfbuzz-7.3.0/build[0m
Build type: [1mcross build[0m
Project name: [1mharfbuzz[0m
Project version: [1m7.3.0[0m
C compiler for the host machine: [1mi686-w64-mingw32-gcc[0m (gcc 12.3.0 "i686-w64-mingw32-gcc (GCC) 12.3.0")
C linker for the host machine: [1mi686-w64-mingw32-gcc[0m ld.bfd 2.40
C++ compiler for the host machine: [1mi686-w64-mingw32-g++[0m (gcc 12.3.0 "i686-w64-mingw32-g++ (GCC) 12.3.0")
C++ linker for the host machine: [1mi686-w64-mingw32-g++[0m ld.bfd 2.40
Compiler for language [1mc[0m for the build machine not found.
Compiler for language [1mcpp[0m for the build machine not found.
Build machine cpu family: [1mx86_64[0m
Build machine cpu: [1mx86_64[0m
Host machine cpu family: [1mx86[0m
Host machine cpu: [1mi686[0m
Target machine cpu family: [1mx86[0m
Target machine cpu: [1mi686[0m
Compiler for C++ supports link arguments -Bsymbolic-functions: [1;32mYES[0m 
Compiler for C++ supports arguments -fno-exceptions: [1;32mYES[0m 
Compiler for C++ supports arguments -fno-rtti: [1;32mYES[0m 
Compiler for C++ supports arguments -fno-threadsafe-statics: [1;32mYES[0m 
Compiler for C++ supports arguments -fvisibility-inlines-hidden: [1;32mYES[0m 
Compiler for C++ supports arguments -Wa,-mbig-obj: [1;32mYES[0m 
Library [1mm[0m found: [1;32mYES[0m
Found pkg-config: [1m/nix/store/gblj7xj9g8j4fma1y7vi4n9dx5q830m8-i686-w64-mingw32-pkg-config-wrapper-0.29.2/bin/i686-w64-mingw32-pkg-config[0m (0.29.2)
Run-time dependency [1mfreetype2[0m found: [1;32mYES[0m [36m26.1.20[0m
Found CMake: [1;31mNO[0m
Run-time dependency [1mglib-2.0[0m found: [1;31mNO[0m (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency [1mglib-2.0[0m

meson.build:114:11: [1;31mERROR:[0m Automatic wrap-based subproject downloading is disabled

A full log can be found at [1m/build/harfbuzz-7.3.0/build/meson-logs/meson-log.txt[0m

```

Reply:
I think it's due to the `lib.optionals (!stdenv.hostPlatform.isMinGW) [ glib ]` part of this PR, which causes Meson not to find `glib` in the last lines of the log. Does `glib` in nixpkgs currently support MinGW?

Reply:
glib currently does not have mingw support 
```
nix build '.#pkgs.pkgsCross.mingw32.glib'
warning: Git tree '/home/sean/nixpkgs-contributing' is dirty
error:
       … in the condition of the assert statement

         at /nix/store/8ff3pyr468q8kjsd4a1b7nxvcfp36anj-source/lib/customisation.nix:263:17:

          262|     in commonAttrs // {
          263|       drvPath = assert condition; drv.drvPath;
             |                 ^
          264|       outPath = assert condition; drv.outPath;

       … while evaluating the attribute 'handled'

         at /nix/store/8ff3pyr468q8kjsd4a1b7nxvcfp36anj-source/pkgs/stdenv/generic/check-meta.nix:499:7:

          498|       # or, alternatively, just output a warning message.
          499|       handled =
             |       ^
          500|         (

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Package ‘glib-2.78.1’ in /nix/store/8ff3pyr468q8kjsd4a1b7nxvcfp36anj-source/pkgs/development/libraries/glib/default.nix:274 is not available on the requested hostPlatform:
         hostPlatform.config = "i686-w64-mingw32"
         package.meta.platforms = [
           "i686-cygwin"
           "x86_64-cygwin"
           "x86_64-darwin"
           "i686-darwin"
           "aarch64-darwin"
           "armv7a-darwin"
           "i686-freebsd13"
           "x86_64-freebsd13"
           "x86_64-solaris"
           "aarch64-linux"
           "armv5tel-linux"
           "armv6l-linux"
           "armv7a-linux"
           "armv7l-linux"
           "i686-linux"
           "loongarch64-linux"
           "m68k-linux"
           "microblaze-linux"
           "microblazeel-linux"
           "mips-linux"
           "mips64-linux"
           "mips64el-linux"
           "mipsel-linux"
           "powerpc64-linux"
           "powerpc64le-linux"
           "riscv32-linux"
           "riscv64-linux"
           "s390-linux"
           "s390x-linux"
           "x86_64-linux"
           "aarch64-netbsd"
           "armv6l-netbsd"
           "armv7a-netbsd"
           "armv7l-netbsd"
           "i686-netbsd"
           "m68k-netbsd"
           "mipsel-netbsd"
           "powerpc-netbsd"
           "riscv32-netbsd"
           "riscv64-netbsd"
           "x86_64-netbsd"
           "i686-openbsd"
           "x86_64-openbsd"
           "x86_64-redox"
         ]
         package.meta.badPlatforms = [ ]
       , refusing to evaluate.

       a) To temporarily allow packages that are unsupported for this system, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnsupportedSystem = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnsupportedSystem = true; }
       to ~/.config/nixpkgs/config.nix.

```

Reply:
In that case, I'll suggest checking [Meson options](https://github.com/harfbuzz/harfbuzz/blob/main/meson_options.txt) to see if `glib` can be disabled cleanly with flags added in `mesonFlags`, to avoid using two build systems conditionally in one package.

Reply:
Managed to get this building using just Meson.